feat(v3.0.0): Introduce Binance WebSocket API, introduce multiplex WebSocket client, bump dependencies, remove some deprecated endpoints

tiagosiebler opened the pull request:
## Summary
This major release brings major upgrades to the Node.js, TypeScript & JavaScript SDK for Binance's REST APIs and WebSockets. The WebSocket API is here - send orders at a lower latency over a dedicated WebSocket connection, without the complexity of managing WebSockets. Use the `WebsocketAPIClient` class to use the WebSocket API like a REST API, with virtually no additional complexity.

## High Level Changes
- Significant upgrades to WebSocket handling & capabilities.
  - Added integration for the **WebSocket API** (send orders via WS, await responses). See #283.
  - **Multiple topics now use a single WebSocket connection.**
  - Breaking changes to "custom logger" method names.
  - Breaking changes to emitted WebsocketClient events:
    - `error` has been renamed to `exception`
    - `reply` has been renamed to `response`
  - WS Response events (after subscribing/unsubscribing to/from topics) are now enriched with the subscription request that was sent to the server.
  - WS heartbeats are disabled automatically if ping/pong frames are not supported in your environment
    - This usually only affects you if you are using the SDK in a frontend environment (e.g. a browser), since binance uses ping/pong control frames.
- Introduced `WebsocketAPIClient`.
  - Use the WebSocket API like a REST API!
  - Looks & behaves like a promise-driven REST API class.
  - All commands:
    - Are sent over the WebSocket API.
    - Are fully typed (request & response).
    - Return a promise, so you can use promises/await to wait for the WebSocket API response.
- Dependency updates:
  - Upgraded build version to Node LTS.
  - Misc dependency updates:
    - Axios to 1.7.9
    - @types/node to LTS
    - typescript to 5.7.3
- Added convenient flag to enable HTTP keepAlive for REST clients.
- Added full support for all available authentication methods on Binance: HMAC, RSA & Ed25519.
- Added support for the Web Crypto API for request sign, both for REST & WS.
  - Node crypto is still used for request sign by default, since Ed25519 (required for WS API) is not stable yet.
- Add `customSignMessageFn()` parameter to WebsocketClient and REST clients.

## Major Changes
##### WebsocketAPIClient
- Introduced a new `WebsocketAPIClient` class, wrapped around the `WebsocketClient`'s WS API capabilities.
- Similar to the REST API clients, this class provides one method for every available WebSocket API command.
- Similar to the REST APIs, commands are wrapped in promises, allowing you to await WebSocket API responses with minimal complexity. The SDK handles the complexity, so for you it feels like a REST API.

##### Multiplex WebsocketClient
- Introduced a newly upgraded `WebsocketClient`, designed around multiplex WebSocket subscriptions.
  - When subscribing to multiple topics within the same product group, they will automatically share one connection for all topics supported by that connection. This is significantly faster than the legacy WS Client, since you will no longer need thousands of connections for thousands of topics.
  - If a topic requires a different connection (e.g. a different subdomain), it will automatically handle all connectivity and event routing. 
  - Detailed usage examples can be seen here: https://github.com/tiagosiebler/binance/blob/wsapi/examples/ws-public.ts#L71
- Full support for Binance's WebSocket API capabilities. Detailed examples can be found in the examples folder & readme.
- For backwards compatibility:
  - Existing "subscribe*" methods are available that exactly match the event behaviour seen in the legacy WS client.
  - Events behave and look the same, thanks to minimal internal processing in how events are received and handled.

##### WebsocketClientV1
- The existing WebsocketClient from the previous release has been renamed to WebsocketClientV1.
  - This should be considered the "legacy" WebsocketClient and may be deprecated in the next major release.
  - It still behaves as before, dedicating one WebSocket connection for every subscribed topic.
  - Detailed behaviour can be seen in the following example: https://github.com/tiagosiebler/binance/blob/wsapi/examples/ws-public-old.ts
  
## Breaking / Potentially Breaking Changes
- WebSockets:
  - For the few places that somewhat supported testnet, renamed the `{ useTestnet: true }` config to `{ testnet: true }`
  - Previous `WebsocketClient` has been renamed to `WebsocketClientV1.`
    - You can keep using this if you're not ready to transition, until you're ready.
    - See below for details.
  - Introduced NEW multiplex WebsocketClient:
    - Capable of live subscribing/unsubscribing (on shared websockets)
    - Supports all userData streams available on Binance.
    - Supports all available WS APIs available on Binance.
  - Consolidated the internal `DefaultLogger` methods, in line with my other SDKs
    - Into 3 key levels: `trace` | `info` | `error`.
    - Similar to the `silly` level before this release, the `trace` level is disabled/silent by default.
  - The "wsKey" will no longer uniquely identify the topic an event is for, since it represents the connection the event came from (and one connection may have multiple subscribed topics).
  - Within the on('open') ws event handler, any references to `event.ws.target.url` should use `event.wsUrl` instead.
- REST Clients:
  - For the few places that somewhat supported testnet, renamed the `{ useTestnet: true }` config to `{ testnet: true }`
  - Deprecated 3rd boolean parameter for testnet. Use `{ testnet: true }` as part of the REST client options instead.
    - Affects USDMClient
    - Affects COINMClient
  - Removed deprecated endpoints
    - `get24hrChangeStatististics()`, use `get24hrChangeStatistics()` instead.
    - `getAccountComissionRate()`, use `getAccountCommissionRate()` instead.
    - `updateAutoInvestmentPlanOld()`, use `updateAutoInvestmentPlan()` instead.
    - `submitAutoInvestmentPlanOld()`, use `submitAutoInvestmentPlan()` instead.
  - Removed old `SpotClient` export
    - Use `MainClient` instead (it is the same).

### High level checklist
- [x] Announce public beta
- [x] Run final batch HFT stress testing via WS API

UjmaIT commented:
Great news thank you for your work. I was just blocked by not working userdata streams so this allows me to keep working. :)

tiagosiebler commented:
> Great news thank you for your work. I was just blocked by not working userdata streams so this allows me to keep working. :)

Glad to hear it! Might not be fully functional as of the latest commit (was working a few commits back), but take a look here for examples:
https://github.com/tiagosiebler/binance/blob/wsapi/examples/ws-api-promises.ts#L121

Also working on a dedicated client for a more rest-like feel (even more than the send fn gives). Something like this, but definitely not finished yet:
https://github.com/tiagosiebler/binance/blob/wsapi/examples/ws-api-client.ts#L57

UjmaIT commented:
Hmm I am having issues with your ws-api-promises.ts Example on testnet with using api key credentials:

Am I doing this correctly ?

`const wsClient = new WebsocketClient(
  {
    api_key: key,
    api_secret: secret,
    beautify: true,
    wsUrl:'wss://stream.testnet.binance.vision:9443' ,
    restOptions:{
        baseUrl:'https://testnet.binance.vision'
    }
  },
  logger, // Optional: inject a custom logger
);
//also updated this line:
const WS_API_WS_KEY: WSAPIWsKey = WS_KEY_MAP.mainWSAPITestnet;
`


But I get this  error in my logs:

`2025-04-06T19:07:12.665Z trace assertIsAuthenticated(): connecting...
2025-04-06T19:07:12.667Z trace assertIsConnected(): connecting...
2025-04-06T19:07:12.668Z trace Opening WS connection to URL: wss://stream.testnet.binance.vision:9443/stream {
  category: "binance-ws",
  wsKey: "mainWSAPITestnet",
}
[ "Websocket connected", {
    category: "binance-ws",
    wsKey: "mainWSAPITestnet",
  } ]
ws connected mainWSAPITestnet
2025-04-06T19:07:13.737Z trace Enabled ping timer {
  category: "binance-ws",
  wsKey: "mainWSAPITestnet",
}
2025-04-06T19:07:13.739Z trace assertIsAuthenticated(): authenticating...
2025-04-06T19:07:13.740Z trace Sending auth request... {
  category: "binance-ws",
  wsKey: "mainWSAPITestnet",
}
2025-04-06T19:07:13.742Z trace assertIsConnected(): connecting...newly connected!
2025-04-06T19:07:13.742Z trace assertIsAuthenticated(): awaiting...
2025-04-06T19:07:13.745Z trace Sending upstream ws message:  {
  category: "binance-ws",
  wsMessage: "{\"id\":1,\"method\":\"session.logon\",\"params\":{\"apiKey\":\"XXXXXX\",\"timestamp\":1743966433741,\"signature\":\"XXXXXXXX=\"}}",
  wsKey: "mainWSAPITestnet",
}
ws exception:  {"error":{"code":2,"msg":"Invalid request: unknown variant `session.logon`, expected one of `SUBSCRIBE`, `UNSUBSCRIBE`, `LIST_SUBSCRIPTIONS`, `SET_PROPERTY`, `GET_PROPERTY`"},"id":1,"wsKey":"mainWSAPITestnet","isWSAPIResponse":true}
[ "Websocket connection closed", {
    category: "binance-ws",
    wsKey: "mainWSAPITestnet",
  } ]
2025-04-06T19:07:14.016Z trace onWsClose(mainWSAPITestnet): rejecting all deferred promises...
[ "Reconnecting to websocket with delay...", {
    category: "binance-ws",
    wsKey: "mainWSAPITestnet",
    connectionDelayMs: 500,
  } ]
`

Is maybe my testnet configuration wrong ?

tiagosiebler commented:
> Hmm I am having issues with your ws-api-promises.ts Example on testnet with using api key credentials:
> 
> Am I doing this correctly ?
> 
> ```
> const wsClient = new WebsocketClient(
>   {
>     api_key: key,
>     api_secret: secret,
>     beautify: true
>   },
>   logger, // Optional: inject a custom logger
> );
> //also updated this line:
> const WS_API_WS_KEY: WSAPIWsKey = WS_KEY_MAP.mainWSAPITestnet;
> ```
> 
> But I get this error in my logs:
> 
> ```
> 2025-04-06T19:07:12.667Z trace assertIsConnected(): connecting...
> 2025-04-06T19:07:12.668Z trace Opening WS connection to URL: wss://testnet.binance.vision/stream {
>   category: "binance-ws",
>   wsKey: "mainWSAPITestnet",
> }
> [ "Websocket connected", {
>     category: "binance-ws",
>     wsKey: "mainWSAPITestnet",
>   } ]
> ws connected mainWSAPITestnet
> 2025-04-06T19:07:13.737Z trace Enabled ping timer {
>   category: "binance-ws",
>   wsKey: "mainWSAPITestnet",
> }
> 2025-04-06T19:07:13.739Z trace assertIsAuthenticated(): authenticating...
> 2025-04-06T19:07:13.740Z trace Sending auth request... {
>   category: "binance-ws",
>   wsKey: "mainWSAPITestnet",
> }
> 2025-04-06T19:07:13.742Z trace assertIsConnected(): connecting...newly connected!
> 2025-04-06T19:07:13.742Z trace assertIsAuthenticated(): awaiting...
> 2025-04-06T19:07:13.745Z trace Sending upstream ws message:  {
>   category: "binance-ws",
>   wsMessage: "{\"id\":1,\"method\":\"session.logon\",\"params\":{\"apiKey\":\"XXXXXX\",\"timestamp\":1743966433741,\"signature\":\"XXXXXXXX=\"}}",
>   wsKey: "mainWSAPITestnet",
> }
> ws exception:  {"error":{"code":2,"msg":"Invalid request: unknown variant `session.logon`, expected one of `SUBSCRIBE`, `UNSUBSCRIBE`, `LIST_SUBSCRIPTIONS`, `SET_PROPERTY`, `GET_PROPERTY`"},"id":1,"wsKey":"mainWSAPITestnet","isWSAPIResponse":true}
> [ "Websocket connection closed", {
>     category: "binance-ws",
>     wsKey: "mainWSAPITestnet",
>   } ]
> 2025-04-06T19:07:14.016Z trace onWsClose(mainWSAPITestnet): rejecting all deferred promises...
> [ "Reconnecting to websocket with delay...", {
>     category: "binance-ws",
>     wsKey: "mainWSAPITestnet",
>     connectionDelayMs: 500,
>   } ]
> ```
> 
> Is maybe my testnet configuration wrong ?

at a glance, looks like it's on my end! I'll try to push a fix today - will tag you.

tiagosiebler commented:
@UjmaIT you can try again with the latest commits on this branch. Testnet connectivity is working for me now with `WS_KEY_MAP.mainWSAPITestnet`. Let me know what other issues you face

tiagosiebler commented:
@UjmaIT for the user data stream, look at this example:
https://github.com/tiagosiebler/binance/blob/wsapi/examples/ws-api-client.ts#L115-L125

I've also implemented a minimal "resubscribe after reconnect" workflow. It's disabled by default but you can enable it using the "onReconnectResubscribeUserDataStream" property shown in the example. Also note that this is with the WebsocketAPIClient class. 

Lastly, keep in mind this is still WIP - many things may still change, but please do let me know thoughts/issues as they arise, as it's the right time to influence how this new functionality evolves.

tiagosiebler commented:
Published binance@3.0.0-beta.4. For anyone trying this out, please be very careful - it's not finished yet.

tiagosiebler commented:
Published binance@3.0.0-beta.5. For anyone trying this out, please be very careful - it's not finished yet.

aweiu commented:
Is it necessary to replace the restAPI related to listenKey with wsAPI? Such as:

- https://github.com/tiagosiebler/binance/blob/wsapi/src/websocket-client.ts#L1001

```
const { listenKey } = await this.wsApiClientCache
        .getSpotWsApiClient(
          this.getWsApiClientOptions(),
          this.options.wsApiOptions,
        )
        .startUserDataStreamForKey(...);
```

- https://github.com/tiagosiebler/binance/blob/wsapi/src/util/websockets/user-data-stream-manager.ts#L433
```
this.wsApiClientCache
          .getSpotWsApiClient(
            this.getWsApiClientOptionsFn(),
            this.getWsClientOptionsfn().WsApiOptions,
          )
          .pingUserDataStreamForKey(listenKey);
```

It seems that Binance also recommends doing this: https://developers.binance.com/docs/binance-spot-api-docs/websocket-api/user-data-stream-requests#listen-key-management-deprecated:

> [!IMPORTANT] These requests have been deprecated, which means we will remove them in the future. Please subscribe to the User Data Stream through the [WebSocket API](https://developers.binance.com/docs/binance-spot-api-docs/websocket-api/user-data-stream-requests#user_data_stream_susbcribe) instead.

aweiu commented:
> https://github.com/tiagosiebler/binance/pull/509#issuecomment-2812328459

I see, it appears that Binance Spot trading is aiming to abandon the entire user data monitoring model based on listenKey, whereas I originally thought that it was only abandoning the restAPI related to listenKey. This is a good thing, I think, as we now only need to subscribe via the `userDataStream.subscribe` websocket API, which bypasses the need to manage the lifecycle of listenKey, and can therefore reduce complexity.

Unfortunately, it seems that only the existing products have implemented the `userDataStream.subscribe` websocket API:
<img width="1074" alt="image" src="https://github.com/user-attachments/assets/59c34ab0-808f-4f8a-a79b-3c665b2cd436" />

tiagosiebler commented:
> > [#509 (comment)](https://github.com/tiagosiebler/binance/pull/509#issuecomment-2812328459)
> 
> I see, it appears that Binance Spot trading is aiming to abandon the entire user data monitoring model based on listenKey, whereas I originally thought that it was only abandoning the restAPI related to listenKey. This is a good thing, I think, as we now only need to subscribe via the `userDataStream.subscribe` websocket API, which bypasses the need to manage the lifecycle of listenKey, and can therefore reduce complexity.
> 
> Unfortunately, it seems that only the existing products have implemented the `userDataStream.subscribe` websocket API:

Yes, it'll make the user data stream simpler for spot markets. Unfortunately also yes, it's only available without the listen key for spot right now - I have asked about the state for other product groups is in the meantime, and binance have told me only the spot listen key workflow is deprecated, no plans / ETAs yet for other product groups. Either way, once I'm done with this PR, it'll easily support the new method for other product groups once it's available, possibly without even needing a code change in the SDK.

tiagosiebler commented:
Published binance@3.0.0-beta.6. For anyone trying this out, please be very careful - it's not finished yet. Almost, but not yet.

tiagosiebler commented:
Since the previous beta:
- simple & consistent way of enabling testnet, REST & WS (`testnet: true` in the config; breaking change for the few places that kinda supported it, `useTestnet:true` renamed to `testnet:true`)
- added http keep alive for all REST clients (works the same as the bybit SDK)
- docs/readme/examples refactoring etc
- refactoring around wsapi types (req/resp)

tiagosiebler commented:
Published binance@3.0.0-beta.7. For anyone trying this out, please be very careful - it's not finished yet. Almost, but not yet.

Since previous beta:
- Implemented injectable customSignMessageFn for all REST clients

JJ-Cro commented on examples/README.md:
````diff
@@ -1,23 +1,95 @@
+Samples that refer to API credentials using `process.env.API_KEY_COM` can be spawned with environment variables. Unix/macOS example:
 ```
-APIKEY="apikeypastedhere" APISECRET="apisecretpastedhere" ts-node examples/ws-userdata.ts
+APIKEY="apikeypastedhere" APISECRET="apisecretpastedhere" tsx examples/WebSockets/ws-userdata-listenkey.ts
````
Here,  lets use single quotes instead of double. Just to make sure if anyone ever does anything, even though there is no passphrase here, you never know what might be problematic later.

JJ-Cro commented on examples/WebSockets/ws-public.ts:
```diff
@@ -0,0 +1,357 @@
+
+  // Formatted data that has gone through the beautifier
+  wsClient.on('formattedMessage', (data) => {
+    if (isWsAggTradeFormatted(data)) {
```
Do you want to keep all of these? Might comment some out, because like this it is confusing. I dont think people need these checks, it was just for internal check to see if the guards work? 

I would keep them out.

tiagosiebler replied:
Agree, might be clearer to keep it commented out with more info on what they're for

JJ-Cro commented on examples/auth/rest-private-ed25519.md:
```diff
@@ -0,0 +1,69 @@
+
+## Creating Ed25519 Keys
+
+Officially, binance recommends downloading and running a key generator from their repo. Guidance for this can be found on the binance website when trying to add a new Ed25519 API key, or in their GitHub repository: https://github.com/binance/asymmetric-key-generator
```
Can add a link to binance guide here.

tiagosiebler replied:
Which link do you mean?

JJ-Cro replied:
"Guidance for this can be found on the binance website when trying to add a new Ed25519 API key"

JJ-Cro commented on examples/misc/rest-spot-exchange-info.ts:
```diff
@@ -85,42 +89,6 @@ async function getSymbolInfo(
   }
 }
 
-/**
```
Why was this part removed? That was kinda the point of the exchange info example? Just wondering.

JJ-Cro replied:
or you moved it somewhere else ?

tiagosiebler replied:
Since they're pure functions, I've exported them as part of the SDK, so anyone can use them in their own projects:
https://github.com/tiagosiebler/binance/blob/wsapi/src/util/rounding.ts

tiagosiebler commented:
Thanks for opening those issues @davydof - I've fixed #551 in the latest commit (beta release pending), and commented on the other one. Let me know if you notice any other discrepancies!

aweiu commented:
```
const ws = new WebsocketClient();
ws.subscribeSymbolBookTicker('btcusdt', 'spot')

// ws.subscribe('btcusdt@bookTicker', 'main')
```

when subscribe `spot` SymbolBookTicker, it will log: 

```
[
  '!!!! Unhandled non-string event type "undefined". Defaulting to "update" channel...',
  {
    u: 68221184661,
    s: 'BTCUSDT',
    b: '93866.50000000',
    B: '0.97083000',
    a: '93866.51000000',
    A: '6.01452000'
  }
]
```

tiagosiebler commented:
> ```
> const ws = new WebsocketClient();
> ws.subscribeSymbolBookTicker('btcusdt', 'spot')
> 
> // ws.subscribe('btcusdt@bookTicker', 'main')
> ```
> 
> when subscribe `spot` SymbolBookTicker, it will log:
> 
> ```
> [
>   '!!!! Unhandled non-string event type "undefined". Defaulting to "update" channel...',
>   {
>     u: 68221184661,
>     s: 'BTCUSDT',
>     b: '93866.50000000',
>     B: '0.97083000',
>     a: '93866.51000000',
>     A: '6.01452000'
>   }
> ]
> ```

Helpful, thanks! This is something I've been meaning to rework... working on an improved parsing workflow (which leads to this issue right now)

tiagosiebler commented:
Published binance@3.0.0-beta.11. For anyone trying this out, please be very careful - it's not finished yet. Almost, but not yet.

- Fixed top-level parsing/handling/enrichment for incoming events
- Fixed wsMarket & streamName enrichment for incoming events

@aweiu your issue is fixed. Please let me know if you notice anything else! Your feedback has been really helpful!

aweiu commented:
> Published binance@3.0.0-beta.11. For anyone trying this out, please be very careful - it's not finished yet. Almost, but not yet.
> 
> * Fixed top-level parsing/handling/enrichment for incoming events
> * Fixed wsMarket & streamName enrichment for incoming events
> 
> @aweiu your issue is fixed. Please let me know if you notice anything else! Your feedback has been really helpful!

Verified, the issue has been fixed. I also found that `streamName` has been added back, which is also very useful for me!

aweiu commented:
Is it possible to make the `timestamp` (in all private wsAPI payloads) optional?
Like: https://github.com/binance/binance-connector-js/blob/master/common/src/websocket.ts#L738

```
const wsAPI = WebsocketAPIClient(...)
// Current: wsAPI.getFuturesAccountStatus("usdm", { timestamp: Date.now() })
// Except:
wsAPI.getFuturesAccountStatus("usdm", {})
wsAPI.getFuturesAccountStatus("usdm")
```

tiagosiebler commented:
> Is it possible to make the `timestamp` (in all private wsAPI payloads) optional? Like: https://github.com/binance/binance-connector-js/blob/master/common/src/websocket.ts#L738

I genuinely wanted to (and had it like that at first), but then I saw that some command throw an error about an unexpected parameter, if I just blindly include it with every request. Which means either I keep a record of all commands that require it (risks being outdated when new commands are added, until the next PR), or I align with the docs (require the user to provide it as one of the required params, as it says in the docs). If I remember correctly, the exchange info command was one of the ones I noticed throws an error if you send a timestamp.

aweiu commented:
> > Is it possible to make the `timestamp` (in all private wsAPI payloads) optional? Like: https://github.com/binance/binance-connector-js/blob/master/common/src/websocket.ts#L738
> 
> I genuinely wanted to (and had it like that at first), but then I saw that some command throw an error about an unexpected parameter, if I just blindly include it with every request. Which means either I keep a record of all commands that require it (risks being outdated when new commands are added, until the next PR), or I align with the docs (require the user to provide it as one of the required params, as it says in the docs). If I remember correctly, the exchange info command was one of the ones I noticed throws an error if you send a timestamp.

Got it! It is also important to ensure that the wsAPI input parameters are consistent with those passed in by the user.

aweiu commented:
### Why are all UserDataStream subscription methods designed with a wsKey parameter? such as `subscribeSpotUserDataStream`、`subscribeUsdFuturesUserDataStream`...

```
const ws = new WebsocketClient(...)
ws.subscribeSpotUserDataStream('usdm')
```

- If you see code like `subscribeSpotUserDataStream('usdm')`, it will be confusing: is it subscribing to `spot` or `usdm`?
- Although wsKey is optional, it seems inconsistent with the design of other methods such as `subscribeSpotSymbolMini24hrTicker`, which do not require a wsKey parameter (after all, wsKey is actually included in the method name)

If the userData subscription method really needs to pass in wsKey, then I think it should be like this: `subscribeUserDataStream(wsKey)`, just like `closeUserDataStream(wsKey)`, developers can easily understand that they are a pair of methods.

If you want to keep all the split subscription methods, should you also add an equal number of unsubscribe methods based on the same logic?Such as:

```
const ws = new WebsocketClient(...)

ws.subscribeSpotUserDataStream()
ws.closeSpotUserDataStream()

ws.subscribeCoinFuturesUserDataStream()
ws.closeCoinFuturesUserDataStream()

// ...
```

tiagosiebler commented:
> ### Why are all UserDataStream subscription methods designed with a wsKey parameter? such as `subscribeSpotUserDataStream`、`subscribeUsdFuturesUserDataStream`...
> ```
> const ws = new WebsocketClient(...)
> ws.subscribeSpotUserDataStream('usdm')
> ```
> 
> * If you see code like `subscribeSpotUserDataStream('usdm')`, it will be confusing: is it subscribing to `spot` or `usdm`?
> * Although wsKey is optional, it seems inconsistent with the design of other methods such as `subscribeSpotSymbolMini24hrTicker`, which do not require a wsKey parameter (after all, wsKey is actually included in the method name)
> 
> If the userData subscription method really needs to pass in wsKey, then I think it should be like this: `subscribeUserDataStream(wsKey)`, just like `closeUserDataStream(wsKey)`, developers can easily understand that they are a pair of methods.

I can understand the confusion here, although you shouldn't set the wsKey parameter unless you specifically want to use one of the alternative wss urls. That's the primary reason I kept a way to inject a non-default wsKey. Spot has multiple, for example:
- Doc: https://developers.binance.com/docs/binance-spot-api-docs/web-socket-streams
- WsKey definitions: https://github.com/tiagosiebler/binance/blob/wsapi/src/util/websockets/websocket-util.ts#L124-L127

Either way, if you don't wish to use an alternative wss url, just call this function without the wsKey parameter. It will default to the wsKey that this function was intended for. If you do for some reason set an incompatible wsKey, it will eventually lead to an error since it will (per your example: usdm on spot user data stream) fetch a listenKey for spot and then try to open a listen key connection for futures -> error with invalid listen key. I've added a comment to each method that'll hopefully make that more obvious for anyone that does get caught out.

> If you want to keep all the split subscription methods, should you also add an equal number of unsubscribe methods based on the same logic?Such as:
> 
> ```
> const ws = new WebsocketClient(...)
> 
> ws.subscribeSpotUserDataStream()
> ws.closeSpotUserDataStream()
> 
> ws.subscribeCoinFuturesUserDataStream()
> ws.closeCoinFuturesUserDataStream()
> 
> // ...
> ```

This is a good point, thanks! I did introduce a close method, although for some reason I defaulted the wsKey to usdm - which isn't intuitive. I'll remove the default and I'll add a matching method for each subscribe method. Thanks for this one!

tiagosiebler commented:
Published binance@3.0.0-beta.14:

- Removed invalid WS API parameters for new spot order (fixes #558). 
- Fixed state cleanup after intentional ws.close and ws.closeAll in the WS client (fixes #560). 

Thank you to everyone involved in testing & feedback, it's been extremely helpful! If you notice anything else that doesn't look normal, please keep opening issues.

tiagosiebler commented:
Published binance@3.0.0-beta.15:
- Fixed intermittent sign failures for REST API calls made with Ed25519 keys.

Thank you to everyone involved in testing & feedback, it's been extremely helpful! If you notice anything else that doesn't look normal, please keep opening issues.

aweiu commented:
```
import { WebsocketAPIClient } from "binance";

ws.getSpotSymbolPriceTicker({ symbol: "BTCUSDT" }).then((rs) =>
  console.log(rs.result),
);
```

=>

```
TypeError: Cannot read properties of undefined (reading 'includes')
      at Object.getSignKeyType (/.../binance@3.0.0-beta.11_typescript@5.7.3/node_modules/binance/src/util/webCryptoAPI.ts:50:14)
```

For the public method getSpotSymbolPriceTicker, the signature module should not be executed?🤔
But even when I pass in the key and secret key, I still get an error:

```
 ws exception:  {"id":1,"status":400,"error":{"code":-1022,"msg":"Signature for this request is not valid."},"rateLimits":[{"rateLimitType":"REQUEST_WEIGHT","interval":"MINUTE","intervalNum":1,"limit":6000,"count":4}],"wsMarket":"spot","wsKey":"mainWSAPI","isWSAPIResponse":true}
[
  'Websocket connection closed',
  { category: 'binance-ws', wsKey: 'mainWSAPI' }
]
[
  'Reconnecting to websocket with delay...',
  {
    category: 'binance-ws',
    wsKey: 'mainWSAPI',
    connectionDelayMs: 500
  }
]
2025-05-16T10:28:26.298Z ws automatically reconnecting....  mainWSAPI
[
  'Reconnecting to websocket now',
  { category: 'binance-ws', wsKey: 'mainWSAPI' }
]
[
  'Websocket reconnected',
  { category: 'binance-ws', wsKey: 'mainWSAPI' }
]
2025-05-16T10:28:27.500Z ws has reconnected  mainWSAPI
```

1. `Signature for this request is not valid`
2. Should this kind of business-level error not trigger a websocketAPI reconnection?

tiagosiebler commented:
@aweiu you're using the WebsocketAPIClient, which is intended for interacting with the WebSocket API in a way similar to a REST API. Think of it like the REST API clients, but auth is always required for it (and happens immediately after opening a connection). If you're only looking to consume events, you should be using the WebsocketClient as before.

aweiu commented:
> @aweiu you're using the WebsocketAPIClient, which is intended for interacting with the WebSocket API in a way similar to a REST API. Think of it like the REST API clients, but auth is always required for it (and happens immediately after opening a connection). If you're only looking to consume events, you should be using the WebsocketClient as before.

However, when I used the websocket API of the [@binance/spot](https://github.com/binance/binance-connector-js/tree/master/clients/spot) library before, if I requested public data, the auth is not required，why？🤔

```
import { Spot } from "@binance/spot";

// @ts-ignore
const client = new Spot({ configurationWebsocketAPI: {}, });

client.websocketAPI
  .connect()
  .then((connection) => {
    return connection.tickerPrice({ symbol: "BTCUSDT" });
  })
  .then((rs) => {
    console.log(rs.data);
  });
```

tiagosiebler commented:
> However, when I used the websocket API of the [@binance/spot](https://github.com/binance/binance-connector-js/tree/master/clients/spot) library before, if I requested public data, the auth is not required，why？🤔
> 
> ```
> import { Spot } from "@binance/spot";
> 
> // @ts-ignore
> const client = new Spot({ configurationWebsocketAPI: {}, });
> 
> client.websocketAPI
>   .connect()
>   .then((connection) => {
>     return connection.tickerPrice({ symbol: "BTCUSDT" });
>   })
>   .then((rs) => {
>     console.log(rs.data);
>   });
> ```

I was largely thinking about account-level commands (submitting orders, changing leverage, etc), which will always require auth, naturally. Part of the motivation for using the WS API is latency, and naturally for market data if latency is important you should be consuming events (as a subscriber), not polling data at intervals (even if the ws api makes it faster than pure REST API).

I also misunderstood what you were doing when I first read your question. I thought you were subscribing to market data as a consumer (which should be via the WebsocketClient). Perhaps auth should remain optional for WebSocket API commands that don't require it (market data). Should be straight forward to implement around how the WebsocketAPIClient sends each command... I'll look into it. In the meantime I've also improved internal checks to make sure there's a clearer error message if API keys are missing when trying to authenticate the WS API.

tiagosiebler commented:
Published binance@3.0.0-beta.16:
- Added flag for public WSAPI commands that do not require auth. Thanks @aweiu for pointing that out.
- Added better error message when API keys are required by WS API but missing.

Thank you to everyone involved in testing & feedback, it's been extremely helpful! If you notice anything else that doesn't look normal, please keep opening issues.

aweiu commented:
What is the difference between the two methods of the WebsocketClient class: `connect` and `connectWSAPI`? At present, I have noticed that the method comments of `connectWSAPI` have stated that it can be called before calling wsAPI, but testing has found that:

1. `ConnectWSAPI` always requires `api_key` and `api_secret`
2. It always fails when `api_key` and `api_secret` are passed in:

```
ws exception:  {"id":1,"status":401,"error":{"code":-2015,"msg":"Invalid API-key, IP, or permissions for action."},"rateLimits":[{"rateLimitType":"REQUEST_WEIGHT","interval":"MINUTE","intervalNum":1,"limit":6000,"count":4}],"wsMarket":"spot","wsKey":"mainWSAPI","isWSAPIResponse":true}
```

I am currently calling `connect` and there are no issues mentioned above (referring to the internal implementation, it seems that all calls `connect`)

tiagosiebler commented:
> What is the difference between the two methods of the WebsocketClient class: `connect` and `connectWSAPI`? At present, I have noticed that the method comments of `connectWSAPI` have stated that it can be called before calling wsAPI, but testing has found that:
> 
> 1. `ConnectWSAPI` always requires `api_key` and `api_secret`
> 2. It always fails when `api_key` and `api_secret` are passed in:
> 
> ```
> ws exception:  {"id":1,"status":401,"error":{"code":-2015,"msg":"Invalid API-key, IP, or permissions for action."},"rateLimits":[{"rateLimitType":"REQUEST_WEIGHT","interval":"MINUTE","intervalNum":1,"limit":6000,"count":4}],"wsMarket":"spot","wsKey":"mainWSAPI","isWSAPIResponse":true}
> ```
> 
> I am currently calling `connect` and there are no issues mentioned above (referring to the internal implementation, it seems that all calls `connect`)

This `connectWSAPI()` was originally built around the assumption that the WS API is for account-level functionality, thus auth is always required. Not the case here, so a question how to anticipate that without diverging from similar logic on other exchanges. I could keep it as is by default but expose an optional boolean? Thoughts?
```typescript
  public connectWSAPI(wsKey: WSAPIWsKey, skipAuth?: boolean): Promise<unknown> {
    if (skipAuth) {
      return this.assertIsConnected(wsKey);
    }

    /** This call automatically ensures the connection is active AND authenticated before resolving */
    return this.assertIsAuthenticated(wsKey);
  }
```

Regarding the difference, as you can see here, the connectWSAPI call is wrapped around the assert methods. Those methods are special in that they're promise-wrapped, although at this point the `assertIsConnected()` call is almost the same as the `connect()` call you're referring to - it resolves a promise when the connection is confirmed to be open (including if it was already opened before that call).

aweiu commented:
> > I have two questions about the wsAPI execution log above:
> > ...

> I've tried this on my end and I don't see any exceptions from this, are you on the latest beta? 

You may misunderstood me:

> **This error was actually caused by me intentionally manually calling wsClient.sendWSAPIRequest (...'time'...) without setting {authIsOptional: true}. Normally, it should not cause the ws connection to be disconnected (currently tested, it seems that as long as the log outputs ws exception the ws connection will be disconnected)**

```
wsClient.sendWSAPIRequest (WS_API_WS_KEY, 'time') // Intentionally not passing in the {authIsOptional: true} parameter
```

** So, my question is actually why a failed request from a certain ws_api leads to the disconnection of the ws connection?**

aweiu commented:
the `keepAlive` option may override `httpsAgent`: https://github.com/tiagosiebler/binance/blob/wsapi/src/util/BaseRestClient.ts#L83

I found this by accident while looking through the binance-connector-js codes: https://github.com/binance/binance-connector-js/blob/master/common/src/utils.ts#L294

tiagosiebler commented:
Hey @aweiu apologies for the delay, had some time away. Thanks for pointing that out. Yes, it was meant initially as a convenience flag for those that didn't have more advanced needs, while it was already possible to inject your own https agent in the constructor of the REST clients (second parameter, passed through to axios). Might be a bit misleading though that it wasn't documented and I've been meaning to improve it. 

I've revised the logic a bit to consider an existing https agent. Also added a test to validate that it's working as expected. Just a simple check that certificate pinning correctly rejects the request when using an incorrect public key hash, while keep alive is active (confirming that both agent configurations were merged, not overwritten). Thanks for all the feedback! 

Let me know if there's anything else! Will look to release this branch this week, since it seems most of the important issues have been ironed out!
